### PR TITLE
[Feature][Kernel] Support INT8 C8 RoPE0 on A2/A3 and INT8/FP8 on A5

### PR DESCRIPTION
**Pull request by Foriv**
### What this PR does / why we need it?

Support `rope_head_dim=0` in the existing C8 `KvQuantSparseFlashAttention` operator:

- A2/A3: INT8
- A5: INT8 and `FLOAT8_E4M3FN`

Callers can provide 512-element queries and 528-byte packed KV rows without an external RoPE payload. Existing RoPE64 inputs remain 576-element queries and 656-byte packed KV rows.

- Reuse the existing Torch/ACLNN interfaces and quantization modes. The default RoPE dimension remains 64.
- Derive GM strides and scale offsets from the actual input width, and zero the absent RoPE lanes inside the retained 576-wide computation buffers.
- Preserve the existing NoPE512 layout, four FP32 per-tile scales and 128-element quantization tile.
- Add A5 `FLOAT8_E4M3FN` RoPE0 host validation while retaining INT8 RoPE0 and the existing HiFloat8 RoPE64 path.
- Include the A5 build-list and shared-buffer/header compatibility changes required by the existing arch35 implementation.
- Keep model-side changes inside the existing SFA call path for empty-RoPE preprocessing, packed cache writes and custom QSFA dispatch. No A5-specific model entrypoint or loading path is added; the C8 path is reached only when SFA is selected and `enable_sparse_sfa_c8=True`. Context-parallel implementations are unchanged.
- Add direct operator accuracy, compact-versus-zero-padded equivalence, graph replay and invalid-input tests, focused Python UTs, and two real A2 RMSNorm/INT8 quantization/cache/attention pipeline tests.

### Does this PR introduce _any_ user-facing change?

Yes. C8 NoPE accepts Q512 and KV528. Each packed KV row contains 512 one-byte NoPE values followed by four FP32 scales at byte offset 512. The one-byte values are INT8 on A2/A3 and INT8 or `FLOAT8_E4M3FN` on A5. Output width remains 512 and the caller-provided attention scale is preserved.

The existing C8 contract remains `key_quant_mode=value_quant_mode=2`, `quant_scale_repo_mode=1` and `tile_size=128`. A2/A3 retain their existing LSE return contract. A5 uses `sparse_block_size=1` and `return_softmax_lse=False`. HiFloat8 remains limited to RoPE64. Existing nonzero-RoPE behavior remains supported. There is no performance-improvement claim.

For the existing A2/A3 model caller, model-config dimensions must match the layer dimensions. This PR does not adapt mixed global-RoPE64/layer-RoPE0 cache allocation. No dedicated context-parallel rejection is added; the existing context-parallel flow is retained. Real multi-card C8 NoPE execution has not been validated.

### How was this patch tested?

Submitted commit: `0956c7700b33cdcbca787c947b795480220c147b`, one signed-off commit.

- A2: targeted OPP build and installation passed, and **54 NPU tests passed**. NPU coverage includes independent accuracy, compact-versus-zero-padded equivalence, LSE, graph replay, invalid inputs, and two real preprocessing/cache/attention pipeline cases. The final cleanup removes only three CPU cases tied to the deleted hardware-adaptor guard; it does not change the validated kernel/cache path.
- A5: rebuilt the targeted OPP and native extension from the updated INT8/FP8 source, then ran **67 direct operator tests: 67 passed**. Coverage includes INT8 and `FLOAT8_E4M3FN`, RoPE0 and RoPE64 equivalence, accuracy, graph replay and host validation.
- `git diff --check` passed on the submitted tree.

Environments: CANN 9.1.0, Python 3.11.10, torch 2.10.0 and torch_npu 2.10.0.post4; A2 910B4-1 and A5 Ascend 950DT.

From the repository root after building and installing the operator for the device:

```bash
# A2/A3
python -m pytest -v tests/e2e/nightly/single_node/ops/singlecard_ops/test_kv_quant_sparse_flash_attention.py tests/e2e/nightly/single_node/ops/singlecard_ops/test_kv_quant_sparse_flash_attention_rope0_a2a3.py

# A5 direct operator tests; the fixture explicitly loads the native extension.
python -m pytest --noconftest -o addopts= -v tests/e2e/nightly/single_node/ops/singlecard_ops/test_kv_quant_sparse_flash_attention_rope0_a5.py
```

Coverage limits: A3 hardware, full-model generation on A2/A5, real multi-card execution, HiFloat8 RoPE0 and performance have not been validated. The two A2 pipeline tests are operator integration tests, not full-model generation.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by xdnjust** [comment]
approve

**Comment by nwpu-zxr** [comment]
The attention backend differences between A5 and A3 are significant. Should a new attention backend be added, or should a new class be created by inheriting the base class to make the changes?

File: csrc/attention/kv_quant_sparse_flash_attention/op_kernel/arch35/kv_quant_sparse_flash_attention_common_arch35.h [hunk 1]
```diff
@@ -67,6 +67,16 @@ enum class QSFATemplateMode {
...
 namespace BaseApi {
+constexpr uint32_t QSFA_NOPE_DIM = 512;
+constexpr uint32_t QSFA_ROPE_DIM = 64;
+constexpr uint32_t QSFA_SCALE_BYTES = 4 * sizeof(float);
+
+__aicore__ inline uint32_t QueryInputDim(const regbaseutil::ConstInfo &info)
+{
+    // Packed KV has one byte per NoPE element and two bytes per RoPE element.
+    return QSFA_NOPE_DIM + (info.dSizeVInput - QSFA_NOPE_DIM - QSFA_SCALE_BYTES) / sizeof(uint16_t);
+}
```
[the remark this answers is not in the document]
Comment (Foriv) [reply csrc/attention/kv_quant_sparse_flash_attention/op_kernel/arch35/kv_quant_sparse_flash_attention_common_arch35.h]: Thanks for flagging this. The value is consumed only after the Host tiling checks have validated the packed MLA contract: `rope_head_dim` must be 0 or 64, and `kHeadDim_` must equal `512 + rope_head_dim * sizeof(uint16_t) + 4 * sizeof(float)`, i.e. 528 or 656 bytes. Therefore `dSizeVInput` is at least `QSFA_NOPE_DIM + QSFA_SCALE_BYTES`, and this expression produces an input dimension of 512 or 576. Returning 512 for an invalid tiling value here would mask a broken Host/Device contract rather than fail it. We will keep the Device calculation tied to the validated tiling contract.

File: csrc/attention/kv_quant_sparse_flash_attention/op_kernel/kv_quant_sparse_flash_attention_kernel_mla.h [hunk 2]
```diff
@@ -216,14 +216,18 @@ template <typename QSFAT> __aicore__ inline void KvQuantSparseFlashAttentionMla<
     constInfo.kvHeadNum = kvHeadNum;
     constInfo.headDim = headDim;
     constInfo.headDimRope = headDimRope;
+    // Packed INT8 KV contains NoPE bytes, optional RoPE, then FP32 scales.
+    constInfo.inputHeadDimRope =
+        (tilingData->baseParams.dSizeVInput * sizeof(KV_T) - headDim * sizeof(KV_T) -
+         headDim / constInfo.tileSize * sizeof(T)) / sizeof(K_ROPE_T);
```
[the remark this answers is not in the document]
Comment (Foriv) [reply csrc/attention/kv_quant_sparse_flash_attention/op_kernel/kv_quant_sparse_flash_attention_kernel_mla.h]: Thanks for the review. This subtraction is downstream of the same Host tiling validation: the NoPE head dimension is 512, `tile_size` is required to be 128, `rope_head_dim` is restricted to 0 or 64, and the packed KV width is checked as `512 + rope_head_dim * 2 + 4 * sizeof(float)` bytes. Consequently the numerator is exactly 0 or 128 bytes and `inputHeadDimRope` is 0 or 64. A ternary fallback here would silently reinterpret invalid tiling data as RoPE0, so the current Device code intentionally relies on the validated Host/Device contract.

File: vllm_ascend/attention/sfa_v1.py [hunk 3]
```diff
@@ -1442,19 +1468,24 @@ def _store_parallel_kv(
             packed_kv = torch.cat(
                 [
                     k_nope.view(-1, k_nope.shape[-1]),
-                    k_pe.view(-1, k_pe.shape[-1]),
+                    k_pe.flatten(0, -2),
                     knope_scale.view(-1, knope_scale.shape[-1]),
```
[the remark this answers is not in the document]
Comment (Foriv) [reply vllm_ascend/attention/sfa_v1.py]: Thanks. Collaborator commit `dda90b0d443ab1556d369d182de3b29efad99d25` addressed the actual RoPE0 issue by changing only the zero-width RoPE component to `k_pe.flatten(0, -2)`, which preserves the leading-row count when the final dimension is zero. `k_nope` comes from `npu_rms_norm`, whose output is contiguous, and the quantization scale is normalized to a contiguous layout before its byte view is returned. Their shape-only `view` calls are intentional and avoid silently copying at the cache-packing boundary.

File: vllm_ascend/attention/sfa_v1.py [hunk 4]
```diff
@@ -1442,19 +1468,24 @@ def _store_parallel_kv(
...
-            )
+            if self.qk_rope_head_dim == 0:
+                # NoPE retains padded input rows during graph replay. Ignore their -1 slots.
+                cache = kv_cache[0]
+                scatter_paged_cache(cache, slot_mapping_sfa[: packed_kv.shape[0]].long(), packed_kv, cache.shape[1])
+            else:
+                torch_npu.npu_scatter_nd_update_(
+                    kv_cache[0].view(-1, packed_head_dim),
+                    slot_mapping_sfa.view(-1, 1),
+                    packed_kv.view(-1, packed_head_dim),
```
[the remark this answers is not in the document]
Comment (Foriv) [reply vllm_ascend/attention/sfa_v1.py]: Thanks. This code path was superseded by collaborator commit `58349017b481f58873f9c3b97a2eb8264a005e0c`: the direct `npu_scatter_nd_update_` call and all three `view` calls in this block were removed. RoPE0 and RoPE64 now both use `DeviceOperator.scatter_cache`, which writes only the actual-token prefix and selects the A5 packed-cache dispatch when applicable.

File: vllm_ascend/attention/sfa_v1.py [hunk 5]
```diff
@@ -1380,9 +1397,13 @@ def _execute_sparse_flash_attention_process(
...
         block_table=None,
     ):
-        if self.qk_rope_head_dim == 0:
+        if self.qk_rope_head_dim == 0 and not self.enable_sparse_sfa_c8:
             return sparse_mla(ql_nope, kv_cache[0], topk_indices, attn_metadata, self.scale)
-        return DeviceOperator.execute_sparse_flash_attention_process(
+        if self.qk_rope_head_dim == 0 and kv_cache[0].shape[1] != attn_metadata.block_size:
+            # Match the logical pages selected by NoPE metadata for hybrid caches.
+            cache = kv_cache[0]
+            kv_cache = (cache.view(-1, attn_metadata.block_size, *cache.shape[2:]), *kv_cache[1:])
```
[the remark this answers is not in the document]
Comment (Foriv) [reply vllm_ascend/attention/sfa_v1.py]: Thanks. This reshape was added by collaborator commit `dda90b0d443ab1556d369d182de3b29efad99d25` to expose the allocator-owned KV cache with the logical page size selected by the NoPE metadata. The cache storage is created as a contiguous tensor, and `view` is intentional here: it is zero-copy and also asserts that storage invariant. `reshape` could allocate a detached copy if that invariant were ever broken, which would be undesirable for a cache passed to the attention operator. We are therefore keeping `view` at this boundary.

**Comment by Foriv** [comment]
@nwpu-zxr We use the same attention backend for A3 and A5. We do not think a new backend or a derived class is necessary, because the existing SFA backend can be reused for both platforms. A larger-scale refactoring can be left to future work.

**Review by chenlongxiao — APPROVED** [review]
